Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
+            CloudBlobContainer container = serviceClient.getContainerReference("libcontainer1");
+            CloudBlockBlob blob = container.getBlockBlobReference(filePathName.getFileName().toString());
+            File downloadingFile = filePathName.toFile();
+            blob.downloadToFile(downloadingFile.getAbsolutePath());
+        } catch (Exception e) {
+            result = false;
+            log.warning("Fail to download file from Azure: " + e.getMessage());
+        }
+
+        return result;
+    }
+
+    /**
+     * Download file and check digest with retry. Return false if all retry fails
+     * @param threadPool
+     * @param filePathNames
+     * @param currentRetry
+     * @param waitTime
+     * @return
+     */
+    private static boolean downloadAndCheckDigest(@NotNull ExecutorService threadPool, @NotNull List<Path> filePathNames, int currentRetry, int waitTime) {
+        List<Path> digestPathNames = new ArrayList<>();
+        for (int i = 0; i < filePathNames.size(); i++) {
+            digestPathNames.add(Paths.get(FilenameUtils.removeExtension(filePathNames.get(i).toString()) + ".md5"));
+        }
+
+        boolean result = true;
+        try {
+            int i = filePathNames.size() - 1;
+            for (; i >= 0; i--) {
+                if (checkFileDigest(filePathNames.get(i), digestPathNames.get(i)) == true)  {
+                    filePathNames.remove(i);
+                    digestPathNames.remove(i);
+                }
+            }
+
+            if (filePathNames.size() == 0) return true;
+
+            if (currentRetry >= MAX_RETRY_TIMES) {
+                log.warning("Maximum retry time reached. Fail to download" + filePathNames.get(0));
+                return false;
+            }
+
+            for (i = 0; i < filePathNames.size(); i++) {
+                FileUtils.deleteQuietly(filePathNames.get(i).toFile());
+                FileUtils.deleteQuietly(digestPathNames.get(i).toFile());
+            }
+
+            List<Callable<Boolean>> downloadThreads = new ArrayList<>();
+            for (i = 0; i < filePathNames.size(); i++) {
+                final Path filePathName = filePathNames.get(i);
+                final Path digestPathName = digestPathNames.get(i);
+                downloadThreads.add(new Callable<Boolean>() {
+                    @Override
+                    public Boolean call() throws Exception {
+                        return downloadFromAzure(filePathName);
+                    }
+                });
+
+                downloadThreads.add(new Callable<Boolean>() {
+                    @Override
+                    public Boolean call() throws Exception {
+                        return downloadFromAzure(digestPathName);
+                    }
+                });
+            };
+
+            List<Future<Boolean>> results = threadPool.invokeAll(downloadThreads, waitTime, TimeUnit.SECONDS);
+
+            for (i = 0; i < results.size(); i++) {
```

Comment:
`for` is evil. 😉

Reply:
Not completely agree:)

---
In reply to: [156555223](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156555223) [](ancestors = 156555223)